src/dev/flang/ast/AbstractType.java

```diff
@@ -995,6 +995,46 @@ public String asString()
     return result;
   }
 
+
+  /**
+   * Check if contraints of this type are satisfied.
+   * Returns itself on success or t_ERROR if constraints are not met.
+   */
+  public AbstractType checkConstraints(SourcePosition pos)
+  {
+    // NYI caching?
+    var result = this;
+    if (!isGenericArgument())
+      {
+        // NYI deduplicate this code?: also in Call.checkTypes()
+
+        // Check that generics match formal generic constraints
+        var fi = featureOfType().generics().list.iterator();
+        var gi = generics().iterator();
+        while (fi.hasNext() &&
+              gi.hasNext()    ) // NYI: handling of open generic arguments
+          {
+            var f = fi.next();
+            var g = gi.next();
+            g.checkConstraints(pos);
+            if(compareTo(f.constraint()) != 0) {
```

Comment:
I think this is missing a `g.`: 
```
  if (g.compareTo(f.constraint()) != 0)
```
Also, pleas add a space between `if` and `(` and place `{` into a new line.

Reply:
@fridis I fixed the indentation. What I want to express here is that the constraint is not the same as the feature itself like in i.e. hasEquals:
```
hasEquals(HE (hasEquals HE).type) is
```

Reply:
@fridis I added a NYI comment regarding the possible infinite recursion